rfc/0000-batch-job-groups.rst

```diff
@@ -0,0 +1,1179 @@
+insert operation. The job group ID is inserted into a table that
+tracks which job groups have been cancelled.
+
+The autoscaler and scheduler avoid trying to spin up resources for
+jobs in job groups that have been cancelled, but the individual job
+has not been cleaned up yet by ignoring any jobs that are in cancelled
+batches or job groups (identical to the current behavior). For an
+accurate fair share computation, the modified
+``user_inst_coll_resources`` table keeps track of the number of ready
+jobs, running jobs, etc. per user, per instance collection, and now
+**per batch**. When a batch has been cancelled or a job group is
+actively being cancelled, then those rows of the table pertaining to
+the specific batch are skipped. This design is a rework of the current
+cancellable resources tables. Because we don't need to track the
+cancellable states of every job group, we can have job groups that
+don't follow a tree like structure and still be able to cancel them
+quickly and not have any performance regressions or incorrect fair
+share computations that affect other user's resource allocations and
+cluster efficiency.
+
+The canceller looks for ready or running jobs in batches that have
+been cancelled or in any job group that has been cancelled and then
+cancels each job one at a time (identical to the current behavior).
+
+Note that because we've added a new field to the
+``user_inst_coll_resources`` table and parameterized it by batch id,
+we'll need to add more garbage collection to remove those rows for
+batches that have been completed (see below). In addition, this design
+means that a cancellation of one job group has temporarily prevented
+the entire batch from being seen by the autoscaler and scheduler. I
+think for the most common use case, this constraint is okay. Most
+batches are small and the QoB use case has all running jobs in the
+same job group so there is no change in behavior from what we
+currently do.
+
+
+~~~~~~~~~~~~~~
+Job Completion
+~~~~~~~~~~~~~~
+
+When a job is marked complete, all job groups the job is a member of
+are checked to see if the number of jobs in the job group is equal to
+the number completed. We are guaranteed that the job that sees the
+number of jobs equals the number completed is the last job to complete
+despite no locking being done. We then execute the callback for any
+newly completed job groups. The amount of extra overhead in the mark
+job complete SQL procedure compared to what we have now is
+O(n_job_groups) the job is a member of, which will need some sort of
+bound on it. This is because we have to update values in the billing
+tables and the table that keeps track of the job states per job group
+for each job group the job is a member of. When the batch is
+completed, we will delete the extra rows from the
+``user_inst_coll_resources`` to make sure that table is as fast as
+possible (O(n_active_batches)).
+
+
+~~~~~~~~~~~~~~~
+Job Group Trees
+~~~~~~~~~~~~~~~
```

Comment:
If this isn't part of the current RFC, let's remove or put in a future work section.